### PR TITLE
Add isInitialized to Message

### PR DESCRIPTION
**Pull request by thomasvl**
I still need to look at pulling over tests for the new method before this is ready to submit, but I wanted to get this up for initial feedback.

### CODE REVIEW

File: Sources/protoc-gen-swift/MessageGenerator.swift [hunk 1]
```diff
@@ -739,3 +944,40 @@ class MessageGenerator {
...
+      if (f.isMessage || f.isGroup) &&
+        HasRequiredFieldsInner(context.getMessageForPath(path: f.typeName)!) {
+        return true
+      }
+    }
+
+    return false
+  }
+
+  let msgDesc = context.getMessageForPath(path: fieldDesc.typeName)!
```
Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: Our use of getMessageForPath suddenly worries me.  [descriptor.proto's](https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.proto#L174-L178) seems to say that string we're using might not be full paths, and could just be the leaf from the parent, but the way we use it it seems like it always has to be full.

File: Sources/protoc-gen-swift/MessageGenerator.swift [hunk 2]
```diff
@@ -718,6 +813,116 @@ class MessageGenerator {
...
+          if f.descriptor.isMessage && HasRequiredFields(f.descriptor, context: context) {
+            p.print("case .\(f.swiftName)(let v):\n")
+            p.indent()
+            p.print("if !v.isInitialized {return false}\n")
+            p.outdent()
+          } else {
+            needsDefault = true
+          }
+        }
+        p.print("case .None:\n")
```
Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: When #133 is done, we can drop the `.None` cases from there, shortening them a hair.

File: Tests/SwiftProtobufTests/unittest.pb.swift [hunk 3]
```diff
@@ -6947,6 +6991,18 @@ struct ProtobufUnittest_TestRequiredOneof: SwiftProtobuf.Message, SwiftProtobuf.
 
     init() {}
 
+    var isInitialized: Bool {
```
Comment (thomasvl) [inline Tests/SwiftProtobufTests/unittest.pb.swift]: Does it make sense to try to optimize this in the generator, and if only one case of the oneof is a message that needs to be checked, instead generate an if to test that one instead of a switch?

**Comment by thomasvl** [comment]
Ok, tests added, and I managed to find a bug. I've left the bug fix and regeneration as a set of commits in the chain so the diffs are more inspectable, but I'll reorder/squash them into the first three commits before we merge this so the history on Master is a little more clean/easier to follow.

File: Sources/protoc-gen-swift/MessageGenerator.swift (hunk 1 above)
Comment (tbkka) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: As far as I can tell, protoc does actually guarantee that these strings are always full paths.  This does not seem to be documented anywhere, though.

Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: Upstreaming the issue (https://github.com/google/protobuf/issues/2711) to hopefully get the definition to say they are always complete.

File: Sources/SwiftProtobuf/Internal.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// -----------------------------------------------------------------------------
+
+
+import Foundation
+import Swift
+
+public struct Internal {
+  private init() {}
+
+  public static func areAllInitialized<M: Message>(_ listOfMessages: [M]) -> Bool {
```
Comment (allevato) [inline Sources/SwiftProtobuf/Internal.swift]: I wonder how different the performance/code size would be if we made this non-generic and the argument type `[Message]` instead (and below).

In this case, since we don't need the generics, we should probably drop them.

File: Sources/SwiftProtobuf/Message.swift [hunk 5]
```diff
@@ -41,6 +41,10 @@ public protocol Message: CustomDebugStringConvertible {
   var anyTypePrefix: String { get }
   var anyTypeURL: String { get }
 
+  // Check if all required fields (if any) have values set on this message
```
Comment (allevato) [inline Sources/SwiftProtobuf/Message.swift]: Use three-slashes so this gets picked up as a doc-comment.

File: Sources/protoc-gen-swift/MessageFieldGenerator.swift [hunk 6]
```diff
@@ -52,6 +52,28 @@ extension Google_Protobuf_FieldDescriptorProto {
         let m = context.getMessageForPath(path: typeName)!
         return m.options.mapEntry
     }
+    // Returns true/false if the field will be hold a Message. map<> is the
+    func getFieldHoldsMessage(context: Context) -> Bool {
+        switch label {
+        case .required:
```
Comment (allevato) [inline Sources/protoc-gen-swift/MessageFieldGenerator.swift]:
```swift
case .required, .optional:
  return type == .message || type == .group
```

File: Sources/protoc-gen-swift/MessageFieldGenerator.swift [hunk 7]
```diff
@@ -52,6 +52,28 @@ extension Google_Protobuf_FieldDescriptorProto {
...
+            fallthrough
+        case .optional:
+            return type == .message || type == .group
+        case .repeated:
+            if type == .group { return true }
```
Comment (allevato) [inline Sources/protoc-gen-swift/MessageFieldGenerator.swift]: The indentation in this block looks like a combination of 2 and 4 spaces, making it a little hard to follow.

File: Sources/protoc-gen-swift/MessageGenerator.swift [hunk 8]
```diff
@@ -718,6 +813,116 @@ class MessageGenerator {
         p.print("}\n")
     }
 
+    func generateIsInitialized(printer p: inout CodePrinter) {
+
```
Comment (allevato) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: Delete extra blank line

File: Sources/protoc-gen-swift/MessageGenerator.swift [hunk 9]
```diff
@@ -739,3 +944,40 @@ class MessageGenerator {
         }
     }
 }
+
+// The logic for this check comes from google/protobuf; the C++ and Java generators specificly.
+//
+// This is a helper for generating isInitialized methods.
+fileprivate func HasRequiredFields(_ fieldDesc: Google_Protobuf_FieldDescriptorProto, context: Context) -> Bool {
```
Comment (allevato) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: Nit: lowercase `hasRequiredFields`

File: Sources/SwiftProtobuf/Internal.swift [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+/// just so the generated code can call them, but shouldn't be called
+/// by developers directly.
+///
+// -----------------------------------------------------------------------------
+
+
+import Foundation
+import Swift
+
+public struct Internal {
```
Comment (thomasvl) [inline Sources/SwiftProtobuf/Internal.swift]: @allevato @tbkka Does it make sense to move any other stuff here?  Especially some of the helper types used in parsing/generating?  I assume they don't have to physically move, just get updated to nest them in an extension of this.

Comment (allevato) [inline Sources/SwiftProtobuf/Internal.swift]: There will be some limits to what we can move—e.g., generic types can't be nested yet, so those would have to stay at the top level. But there's probably a good amount of public-as-implementation-detail stuff that we could put in there and it would be a good idea to do that.

File: Sources/SwiftProtobuf/Message.swift (hunk 5 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/Message.swift]: Done.

File: Sources/SwiftProtobuf/Internal.swift (hunk 4 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/Internal.swift]: Done

File: Sources/protoc-gen-swift/MessageFieldGenerator.swift (hunk 6 above)
Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageFieldGenerator.swift]: Done.

File: Sources/protoc-gen-swift/MessageFieldGenerator.swift (hunk 7 above)
Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageFieldGenerator.swift]: I think I got it all.

File: Sources/protoc-gen-swift/MessageGenerator.swift (hunk 8 above)
Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: Done.

File: Sources/protoc-gen-swift/MessageGenerator.swift (hunk 9 above)
Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageGenerator.swift]: This sucked.  As `swiftc` then got grumpy about the local with the same name, so I went ahead and renamed it changing the calling a little.

File: Sources/protoc-gen-swift/MessageFieldGenerator.swift [hunk 11]
```diff
@@ -52,6 +52,26 @@ extension Google_Protobuf_FieldDescriptorProto {
         let m = context.getMessageForPath(path: typeName)!
         return m.options.mapEntry
     }
+    // Returns true/false if the field will be hold a Message. map<> is the
+    func getFieldHoldsMessage(context: Context) -> Bool {
```
Comment (allevato) [inline Sources/protoc-gen-swift/MessageFieldGenerator.swift]: Is there a reason this is a separate function, instead of putting the body in the property below that calls it? That would get rid of the awkward name.

Comment (thomasvl) [inline Sources/protoc-gen-swift/MessageFieldGenerator.swift]: Just following the general flow between the two, moved.

**Comment by thomasvl** [comment]
Reordered and folded into 4 logic commits for landing.
